### ISSUE TITLE
Add option to always focus certain applications

### ISSUE THREAD

**Issue by dominiklohmann** [koekeishiya/kwm#205]
I need this for Little Snitch, which cannot be focused with kwm active and thus cannot be controlled using keyboard shortcuts. Screenshot with Accessibility Inspector open that shows how Little Snitch is not focused despite it being the frontmost window and the cursor being on it:

<img width="974" alt="1456999854" src="https://cloud.githubusercontent.com/assets/4488655/13491184/439e42b4-e131-11e5-96f5-c35618969355.png">

`kwmc config float Little Snitch` does not help here.

**Comment by koekeishiya** [comment]
Looke like #40

**Comment by dominiklohmann** [comment]
Tried this with both `AXUnknown`, `AXWindow` and `AXStandardWindow` - neither of them works.
What I really need is a way to force kwm to have `Little Snitch Configuration` stay at the top whenever it is active. Trying the same with `Little Snitch` doesn't work either.

```
kwmc config add-role AXUnknown Little Snitch Configuration
kwmc config float Little Snitch Configuration
```

**Comment by koekeishiya** [comment]
> right click the .app folder and click "show package contents".
Open the Info.plist file and locate the string "Bundle Name"

Make sure you pass the proper name. If the method aboves does not work ,there is nothing I can do about it either.

**Comment by dominiklohmann** [comment]
That's what I did, the value for "Bundle Name" is "Little Snitch Configuration". 

I guess I'll leave the issue open for now?

**Comment by koekeishiya** [comment]
Does it appear in the list returned by `kwmc read windows` ?

**Comment by dominiklohmann** [comment]
Nope.

**Comment by koekeishiya** [comment]
Add this to [window.cpp:69](https://github.com/koekeishiya/kwm/blob/master/kwm/window.cpp#L69-L71) and run a debug build from the terminal, on the space that has the Little Snitch window open. Then do ´kwmc read windows` again.

```
else
{
    DEBUG("Role Mismatch: " << KWMTiling.FocusLst[WindowIndex].Owner)
}
```

**Comment by dominiklohmann** [comment]
Monitored both stdout and stderr with tail -f, doesn't show up.

However, the window does get highlighted for a split second when I click on it.

<img width="1440" alt="1457006926" src="https://cloud.githubusercontent.com/assets/4488655/13493914/5a4453e0-e141-11e5-81e0-7660051d775c.png">

**Comment by koekeishiya** [comment]
Try comment out [window.cpp:307](https://github.com/koekeishiya/kwm/blob/master/kwm/window.cpp#L307), this should be the reason.

**Comment by dominiklohmann** [comment]
Yes, that fixed it.

**Comment by koekeishiya** [comment]
> Yes, that fixed it.

Alright, can you print the value of "Window->Layer"

**Comment by dominiklohmann** [comment]
1490 for "Little Snitch", 0 for everything else.

**Comment by koekeishiya** [comment]
No idea what the "CGWindowLevelForKey(kCGPopUpMenuWindowLevelKey);" constant for this ID is, but adding 1490 to the layer values checked in the function "ShouldWindowGainFocus()" should do the trick.

I have no idea if this causes problems for other applications such as popclip, so it would be nice if you can test this change locally, and report back on how it works out.

**Comment by dominiklohmann** [comment]
Addind `Window->Layer == 1490` to the checked Layer IDs did the trick for me.

Do you know if there is a list of all `CGWindowLevelKey` constants? 

> I have no idea if this causes problems for other applications such as popclip, so it would be nice if you can test this change locally, and report back on how it works out.

Me neither, since I don't own any other applications that behave like this. Can someone that owns PopClip or any similar application confirm the behaviour did not change?

**Comment by matthewdias** [comment]
I have popclip. What should I do to test it?

**Comment by dominiklohmann** [comment]
In `kwm/window.cpp`, change from:

```cpp
bool ShouldWindowGainFocus(window_info *Window)
{
    return Window->Layer == CGWindowLevelForKey(kCGNormalWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGFloatingWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGTornOffMenuWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGDockWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGMainMenuWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGMaximumWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGModalPanelWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGUtilityWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGOverlayWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGHelpWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGPopUpMenuWindowLevelKey);
}
```

to:

```cpp
bool ShouldWindowGainFocus(window_info *Window)
{
    return Window->Layer == CGWindowLevelForKey(kCGNormalWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGFloatingWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGTornOffMenuWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGDockWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGMainMenuWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGMaximumWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGModalPanelWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGUtilityWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGOverlayWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGHelpWindowLevelKey) ||
           Window->Layer == CGWindowLevelForKey(kCGPopUpMenuWindowLevelKey) ||
           Window->Layer == 1490;
}
```

After that, make and run it and monitor if you see any behaviour changes, especially with applications like PopClip.

**Comment by matthewdias** [comment]
Seems to work fine with PopClip. I'll keep monitoring but I think this is solved for now.

**Comment by EdenSG** [comment]
I'm having the same issue with the Fantastical 2 menu bar window. It's a normal `AXStandardWindow`, but kwm closes it whenever the cursor moved over it.

**Comment by dominiklohmann** [comment]
Have you tried `kwmc config float Fantastical 2`? Basically I'd try all the things I went through above, and if that does not work, try to find the layer ID of the window.

**Comment by koekeishiya** [comment]
If a window does not get tiled, and does not gain focus, the problem is related to #40
If a window does get tiled, but can not get focused, it's related to the ShouldWindowGainFocus() function and you need to identify the layer of the window.

**Comment by dvcrn** [comment]
`kwmc config float Fantastical 2` worked for me. However fantastical comes with a menu bar version that is always docked. I want kwm to only ignore this window (let it float) but when I open the full fantastical, it should tile it like all the other applications. Is this possible?

<img width="455" alt="2016-08-04 10 55 17" src="https://cloud.githubusercontent.com/assets/688326/17387811/f9356868-5a31-11e6-960f-53fadb353b3f.png">

**Comment by dvcrn** [comment]
Also @dominiklohmann, what were your final settings to deal with little snitch? I am having the same problems you described with kwm 3

**Comment by dominiklohmann** [comment]
I just made it float in my kwmrc, that works for me.

**Comment by chrismetcalf** [comment]
So `kwmc rule owner="Fantastical 2" properties={float="true"}` works for the Fantastical full calendar window, but not for the menu bar pop-up. When I pop that up KWM freaks out and struggles to figure out how to resize it into my panes. 

Any known good rules for that pop-up?
